src/cls/cephfs/cls_cephfs.cc

```diff
@@ -0,0 +1,289 @@
+}
+
+
+/**
+ * Helper for setting an xattr iff the input value compares as
+ * greater than an existing value decoded from the xattr.
+ */
+template <typename A>
+static int set_if_greater(cls_method_context_t hctx, const std::string &xattr_name, const A input_val)
+{
+  bufferlist existing_val_bl;
+
+  bool set_val = false;
+  int r = cls_cxx_getxattr(hctx, xattr_name.c_str(), &existing_val_bl);
+  if (r == -ENOENT || existing_val_bl.length() == 0) {
+    set_val = true;
+  } else if (r == 0) {
+    bufferlist::iterator existing_p = existing_val_bl.begin();
+    try {
+      A existing_val;
+      ::decode(existing_val, existing_p);
+      if (!existing_p.end()) {
+        // Trailing junk?  Consider it invalid and overwrite
+        set_val = true;
+      } else {
+        // Valid existing value, do comparison
+        set_val = input_val > existing_val;
+      }
+    } catch (const buffer::error &err) {
+      // Corrupt or empty existing value, overwrite it
+      set_val = true;
+    }
+  } else {
+    return r;
+  }
+
+  // Conditionally set the new xattr
+  if (set_val) {
+    bufferlist set_bl;
+    ::encode(input_val, set_bl);
+    return cls_cxx_setxattr(hctx, xattr_name.c_str(), &set_bl);
+  } else {
+    return 0;
+  }
+}
+
+static int set_if_greater(cls_method_context_t hctx, const std::string &xattr_name, const ObjCeiling input_val)
+{
+  bufferlist existing_val_bl;
+
+  bool set_val = false;
+  int r = cls_cxx_getxattr(hctx, xattr_name.c_str(), &existing_val_bl);
+  if (r == -ENOENT || existing_val_bl.length() == 0) {
+    set_val = true;
+  } else if (r >= 0) {
+    bufferlist::iterator existing_p = existing_val_bl.begin();
+    try {
+      ObjCeiling existing_val;
+      ::decode(existing_val, existing_p);
+      if (!existing_p.end()) {
+        // Trailing junk?  Consider it invalid and overwrite
+        set_val = true;
+      } else {
+        // Valid existing value, do comparison
+        set_val = input_val > existing_val;
+      }
+    } catch (const buffer::error &err) {
+      // Corrupt or empty existing value, overwrite it
+      set_val = true;
+    }
+  } else {
+    return r;
+  }
+
+  // Conditionally set the new xattr
+  if (set_val) {
+    bufferlist set_bl;
+    ::encode(input_val, set_bl);
+    return cls_cxx_setxattr(hctx, xattr_name.c_str(), &set_bl);
+  } else {
+    return 0;
+  }
+}
+
+static int accumulate_inode_metadata(cls_method_context_t hctx,
+    bufferlist *in, bufferlist *out)
+{
+  assert(in != NULL);
+  assert(out != NULL);
+
+  std::string obj_xattr_name;
+  std::string mtime_xattr_name;
+  uint64_t input_obj_id = 0;
+  uint64_t input_obj_size = 0;
+  time_t input_mtime = 0;
+  int r = 0;
+
+  // Decode `in`
+  bufferlist::iterator q = in->begin();
+  try {
+    ::decode(obj_xattr_name, q);
+    ::decode(mtime_xattr_name, q);
+    ::decode(input_obj_id, q);
+    ::decode(input_obj_size, q);
+    ::decode(input_mtime, q);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
```

Comment:
Rather than doing raw encoding here, you probably want to define a struct which contains the data types, can be properly versioned, and is used for encode/decode on both client and class. See e.g. the src/cls/rgw classes where we have cls_X and cls_X_client files.